### PR TITLE
arm_dynarmic: Coprocessor support

### PR DESCRIPTION
**Pull request by MerryMage**
Been sitting on this for a while.

Documentation of coprocessor interface can be found [here](https://github.com/MerryMage/dynarmic/blob/459d7d1bafcf85677c989b7cb260d3789aa813e2/include/dynarmic/coprocessor.h#L20).

----

There are three possible ways to handle a coprocessor instruction:

* Compile a call to a callback.
* Compile a read/write to a pointer.
* Compile code that raises coprocessor exception.

This PR implements only the parts of cp15 required for Citra.

### CODE REVIEW

File: src/core/arm/dynarmic/arm_dynarmic_cp15.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+#include "core/arm/skyeye_common/arm_regformat.h"
+#include "core/arm/skyeye_common/armstate.h"
+
+using Callback = Dynarmic::Coprocessor::Callback;
+using CallbackOrAccessOneWord = Dynarmic::Coprocessor::CallbackOrAccessOneWord;
+using CallbackOrAccessTwoWords = Dynarmic::Coprocessor::CallbackOrAccessTwoWords;
+
+DynarmicCP15::DynarmicCP15(const std::shared_ptr<ARMul_State>& state) : interpreter_state(state) {}
+
+DynarmicCP15::~DynarmicCP15() {}
```
Comment (lioncash) [inline src/core/arm/dynarmic/arm_dynarmic_cp15.cpp]:
```cpp
DynarmicCP15::~DynarmicCP15() = default;
```

File: src/core/arm/dynarmic/arm_dynarmic_cp15.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <memory>
+#include <dynarmic/coprocessor.h>
+#include "common/common_types.h"
+
+struct ARMul_State;
+
+class DynarmicCP15 final : public Dynarmic::Coprocessor {
+public:
+    DynarmicCP15(const std::shared_ptr<ARMul_State>&);
```
Comment (lioncash) [inline src/core/arm/dynarmic/arm_dynarmic_cp15.h]: This should likely be `explicit`

File: src/core/arm/dynarmic/arm_dynarmic_cp15.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    // TODO(merry): Privileged CP15 registers
+
+    if (!two && CRn == CoprocReg::C7 && opc1 == 0 && CRm == CoprocReg::C5 && opc2 == 4) {
+        return &interpreter_state->CP15[CP15_FLUSH_PREFETCH_BUFFER];
+    }
+
+    if (!two && CRn == CoprocReg::C7 && opc1 == 0 && CRm == CoprocReg::C10) {
+        switch (opc2) {
+        case 4:
+            return &interpreter_state->CP15[CP15_DATA_SYNC_BARRIER];
```
Comment (MerryMage) [inline src/core/arm/dynarmic/arm_dynarmic_cp15.cpp]: Yes, and done.
Perhaps returning a sentinel value (`nullptr`?) that explicitly states these are no-ops could be considered?

**Comment by MerryMage** [comment]
Addressed above.

**Comment by RavenHome1** [comment]
about 10 fps improvement in pokemon  moon intro with bleeding 173 ;)

capture


capture2

**Comment by urdaneta09** [comment]
@RavenHome1   pokemon X got the same boost in intro/in-game 
kingdom hearts dream drop distance only intro, the in-game are the same (slow maybe is more gpu demand i guess)
fire emblem awakening only intro the in-game are the same

**Comment by RavenHome1** [comment]
@urdaneta09 maybe these cpu instructions are heavily used to play intros more than anything.

**Comment by RavenHome1** [comment]
so i noticed citra has been throwing this error randomly in the console ever since this was merged, nothing seems to be breaking but it might something important to bring attention to 

capture

**Pull request merged**